src/V3HierBlock.cpp

```diff
@@ -0,0 +1,436 @@
+//    all parameters must exactly match.
+// 7) In V3HierBlock.cpp, relationship among hierarchical blocks are checked in run a).
+//    (which block uses other blocks..)
+// 8) In V3EmitMk.cpp, ${prefix}_hier.mk is created in run a).
+//
+// There are two hidden command options.
+//   --hierarchical-child is added to Verilator run b).
+//   --hierarchical-block module_name,mangled_name,name0,value0,name1,value1,...
+//       module_name  :The original modulename
+//       mangled_name :Mangled name of parameterized modules (named in V3Param.cpp).
+//                     Same as module_name for non-parameterized hierarchical block.
+//       name         :The name of the parameter
+//       value        :Overridden value of the parameter
+//
+//       Used for b) and c).
+//       This options is repeated for all instantiating hierarchical blocks.
+
+#include <memory>
+#include <sstream>
+#include <utility>
+#include <vector>
+
+#include "V3Ast.h"
+#include "V3Error.h"
+#include "V3File.h"
+#include "V3HierBlock.h"
+#include "V3String.h"
+#include "V3Stats.h"
+
+typedef std::vector<std::pair<string, string> > StrGParams;
+
+static StrGParams stringifyParams(const V3HierBlock::GParams& gparams, bool forGOption) {
+    StrGParams strParams;
+    for (V3HierBlock::GParams::const_iterator gparamIt = gparams.begin();
+         gparamIt != gparams.end(); ++gparamIt) {
+        if (AstConst* constp = VN_CAST((*gparamIt)->valuep(), Const)) {
+            // Only constant parameter needs to be set to -G because already checked in
+            // V3Param.cpp. See also ParamVisitor::checkSupportedParam() in the file.
+            if (constp->isDouble()) {
+                // 64 bit width of hex can be expressed with 16 chars.
+                // 32 chars must be long enough for hexadecial floating point
+                // considering prefix of '0x', '.', and 'P'.
+                std::vector<char> hexFpStr(32, '\0');
+                const int len = VL_SNPRINTF(hexFpStr.data(), hexFpStr.size(), "%a",
+                                            constp->num().toDouble());
+                UASSERT_OBJ(0 < len && static_cast<size_t>(len) < hexFpStr.size(), constp,
+                            " is not properly converted to string");
+                strParams.push_back(std::make_pair((*gparamIt)->name(), hexFpStr.data()));
+            } else if (constp->isString()) {
+                string s = constp->num().toString();
+                if (!forGOption) s = VString::quoteBackslash(s);
+                s = VString::quoteStringLiteralForShell(s);
+                strParams.push_back(std::make_pair((*gparamIt)->name(), s));
+            } else {  // Either signed or unsigned integer.
+                string s = constp->num().ascii(true, true);
+                s = VString::quoteAny(s, '\'', '\\');
+                strParams.push_back(std::make_pair((*gparamIt)->name(), s));
```

Comment:
Suggest move this line out of the 3 if cases.  The if cases would just set "s" appropriately.

Reply:
Done in e03034e5dafb9b461fee8c781bf05156b86bfbc2